File: stepik-union/src/main/java/org/stepik/plugin/refactoring/move/AbstractMoveHandlerDelegate.java
package org.stepik.plugin.refactoring.move;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.ex.MessagesEx;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFileSystemItem;
import com.intellij.refactoring.move.MoveCallback;
import com.intellij.refactoring.move.MoveHandlerDelegate;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepik.plugin.utils.ProjectPsiFilesUtils;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

import static org.stepik.plugin.utils.ProjectPsiFilesUtils.isCanNotBeTarget;
import static org.stepik.plugin.utils.ProjectPsiFilesUtils.isNotMovableOrRenameElement;

/**
 * @author meanmail
 */
public abstract class AbstractMoveHandlerDelegate extends MoveHandlerDelegate {
    private final Set<Class<? extends PsiElement>> acceptableClasses = new HashSet<>();

    protected void addAcceptableClasses(@NotNull Set<Class<? extends PsiElement>> classes) {
        acceptableClasses.addAll(classes);
    }

    @Override
    public boolean isMoveRedundant(PsiElement source, PsiElement target) {
        return isNotMovableOrRenameElement(source, acceptableClasses) || isCanNotBeTarget(target, acceptableClasses);
    }

    @Override
    public boolean canMove(PsiElement[] elements, @Nullable PsiElement targetContainer) {
        return isValidTarget(targetContainer, elements);
    }

    @Override
    public boolean isValidTarget(@Nullable PsiElement target, PsiElement[] sources) {
        for (PsiElement source : sources) {
            if (isNotMovableOrRenameElement(source, acceptableClasses)) {
                return true;
            }
        }

        return isCanNotBeTarget(target, acceptableClasses);
    }

    @Override
    public void doMove(
            Project project,
            PsiElement[] elements,
            @Nullable PsiElement targetContainer,
            @Nullable MoveCallback callback) {
        List<PsiFileSystemItem> sources = Arrays.stream(elements)
                .filter(psiElement -> isNotMovableOrRenameElement(psiElement, acceptableClasses))
                .map(ProjectPsiFilesUtils::getFile)
                .filter(Objects::nonNull)
                .collect(Collectors.toList());

        StringBuilder message = new StringBuilder();
        if (sources.size() > 1) {
            message.append("You can not move the following elements:");
        } else {
            PsiFileSystemItem source = sources.get(0);
            message.append("You can not move the ")
                    .append(source.isDirectory() ? "directory" : "file")
                    .append(":");
        }

        for (PsiFileSystemItem file : sources) {
            message.append("\n")
                    .append(file.getVirtualFile().getPath());
        }

        MessagesEx.error(project, message.toString(), "Move").showNow();
    }
}

File: stepik-union/src/main/java/org/stepik/plugin/refactoring/move/AbstractMoveHandlerDelegate.kt
package org.stepik.plugin.refactoring.move

import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.ex.MessagesEx
import com.intellij.psi.PsiElement
import com.intellij.refactoring.move.MoveCallback
import com.intellij.refactoring.move.MoveHandlerDelegate
import org.stepik.plugin.utils.ProjectPsiFilesUtils
import org.stepik.plugin.utils.ProjectPsiFilesUtils.isCanNotBeTarget
import org.stepik.plugin.utils.ProjectPsiFilesUtils.isNotMovableOrRenameElement
import java.util.*


abstract class AbstractMoveHandlerDelegate : MoveHandlerDelegate() {
    private val acceptableClasses = HashSet<Class<out PsiElement>>()

    protected fun addAcceptableClasses(classes: Set<Class<out PsiElement>>) {
        acceptableClasses.addAll(classes)
    }

    override fun isMoveRedundant(source: PsiElement?, target: PsiElement?): Boolean {
        if (source == null) {
            return true
        }
        return isNotMovableOrRenameElement(source, acceptableClasses) || isCanNotBeTarget(target, acceptableClasses)
    }

    override fun canMove(elements: Array<PsiElement>, targetContainer: PsiElement?): Boolean {
        return isValidTarget(targetContainer, elements)
    }

    override fun isValidTarget(target: PsiElement?, sources: Array<PsiElement>?): Boolean {
        if (sources == null) {
            return false
        }

        return sources.any { isNotMovableOrRenameElement(it, acceptableClasses) } || isCanNotBeTarget(target, acceptableClasses)
    }

    override fun doMove(
            project: Project?,
            elements: Array<PsiElement>?,
            targetContainer: PsiElement?,
            callback: MoveCallback?) {
        val sources = elements!!.filter { isNotMovableOrRenameElement(it, acceptableClasses) }
                .map { ProjectPsiFilesUtils.getFile(it) }
                .filter { Objects.nonNull(it) }

        val message = StringBuilder()
        if (sources.count() > 1) {
            message.append("You can not move the following elements:")
        } else {
            val source = sources.firstOrNull()
            message.append("You can not move the ")
                    .append(if (source?.isDirectory == true) "directory" else "file")
                    .append(":")
        }

        for (file in sources) {
            message.append("\n")
                    .append(file?.virtualFile?.path)
        }

        MessagesEx.error(project, message.toString(), "Move").showNow()
    }
}
